From a0ce6480f38f67584dbc61a294e1178c37bb5f20 Mon Sep 17 00:00:00 2001
From: Rob Evers <revers@redhat.com>
Date: Mon, 23 Jun 2014 21:41:54 -0400
Subject: [scsi] lpfc 10.2.8000.0: Fix locking for postbufq when freeing

Message-id: <1403559726-20488-13-git-send-email-revers@redhat.com>
Patchwork-id: 84421
O-Subject: [RHEL7.1 e-stor PATCH 12/24] scsi: lpfc 10.2.8000.0: Fix locking for postbufq when freeing
Bugzilla: 1088574
RH-Acked-by: Ewan Milne <emilne@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Dick Kennedy <rkennedy@Emulex.Com>

Submitted by James Smart James.Smart@emulex.com

Fix locking for postbufq when freeing

Signed-off-by: James Smart james.smart@emulex.com

http://marc.info/?l=linux-scsi&m=139663390821857&w=2
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/lpfc/lpfc_init.c b/drivers/scsi/lpfc/lpfc_init.c
index 4594f4c..f8ec830 100644
--- a/drivers/scsi/lpfc/lpfc_init.c
+++ b/drivers/scsi/lpfc/lpfc_init.c
@@ -836,19 +836,28 @@ lpfc_hba_free_post_buf(struct lpfc_hba *phba)
  struct lpfc_sli *psli = &phba->sli;
  struct lpfc_sli_ring *pring;
  struct lpfc_dmabuf *mp, *next_mp;
+ LIST_HEAD(buflist);
+ int count;
 
  if (phba->sli3_options & LPFC_SLI3_HBQ_ENABLED)
   lpfc_sli_hbqbuf_free_all(phba);
  else {
   /* Cleanup preposted buffers on the ELS ring */
-  spin_lock_irq(&phba->hbalock);
   pring = &psli->ring[LPFC_ELS_RING];
-  list_for_each_entry_safe(mp, next_mp, &pring->postbufq, list) {
+  spin_lock_irq(&phba->hbalock);
+  list_splice_init(&pring->postbufq, &buflist);
+  spin_unlock_irq(&phba->hbalock);
+
+  count = 0;
+  list_for_each_entry_safe(mp, next_mp, &buflist, list) {
    list_del(&mp->list);
-   pring->postbufq_cnt--;
+   count++;
    lpfc_mbuf_free(phba, mp->virt, mp->phys);
    kfree(mp);
   }
+
+  spin_lock_irq(&phba->hbalock);
+  pring->postbufq_cnt -= count;
   spin_unlock_irq(&phba->hbalock);
  }
 }
-- 
1.7.1